ide/lsp.client/src/org/netbeans/modules/lsp/client/bindings/Formatter.java

```diff
@@ -0,0 +1,153 @@
+import javax.swing.text.Document;
+import org.eclipse.lsp4j.DocumentFormattingParams;
+import org.eclipse.lsp4j.DocumentRangeFormattingParams;
+import org.eclipse.lsp4j.FormattingOptions;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextDocumentIdentifier;
+import org.eclipse.lsp4j.TextEdit;
+import org.netbeans.api.editor.mimelookup.MimeRegistration;
+import org.netbeans.modules.editor.indent.api.IndentUtils;
+import org.netbeans.modules.editor.indent.spi.Context;
+import org.netbeans.modules.editor.indent.spi.ExtraLock;
+import org.netbeans.modules.editor.indent.spi.ReformatTask;
+import org.netbeans.modules.lsp.client.LSPBindings;
+import org.netbeans.modules.lsp.client.Utils;
+import org.openide.filesystems.FileObject;
+import org.openide.loaders.MultiDataObject;
+
+public class Formatter implements ReformatTask {
+
+    private static final Logger LOG = Logger.getLogger(Formatter.class.getName());
+
+    @MimeRegistration(mimeType = "", service = ReformatTask.Factory.class)
+    public static class Factory implements ReformatTask.Factory {
+
+        @Override
+        public ReformatTask createTask(Context context) {
+            return new Formatter(context);
+        }
+
+    }
+
+    private final Context ctx;
+
+    public Formatter(Context ctx) {
+        this.ctx = ctx;
+    }
+
+    @Override
+    public void reformat() throws BadLocationException {
+        Object stream = ctx.document().getProperty(Document.StreamDescriptionProperty);
+        if(stream instanceof MultiDataObject) {
+            MultiDataObject mdo = ((MultiDataObject) stream);
+            FileObject fo = mdo.getPrimaryFile();
+            LSPBindings bindings = LSPBindings.getBindings(fo);
+            if (bindings != null) {
+                Boolean documentFormatting = bindings.getInitResult().getCapabilities().getDocumentFormattingProvider();
+                Boolean rangeFormatting = bindings.getInitResult().getCapabilities().getDocumentRangeFormattingProvider();
+                if (rangeFormatting != null && rangeFormatting) {
+                    rangeFormat(fo, bindings);
+                } else if (documentFormatting != null && documentFormatting) {
+                    documentFormat(fo, bindings);
+                }
+            }
+        }
+    }
+
+    private void rangeFormat(FileObject fo, LSPBindings bindings) throws BadLocationException {
+        DocumentRangeFormattingParams drfp = new DocumentRangeFormattingParams();
+        drfp.setTextDocument(new TextDocumentIdentifier(Utils.toURI(fo)));
+        drfp.setOptions(new FormattingOptions(
+            IndentUtils.indentLevelSize(ctx.document()),
+            IndentUtils.isExpandTabs(ctx.document())));
+        drfp.setRange(new Range(
+            Utils.createPosition(ctx.document(), ctx.startOffset()),
+            Utils.createPosition(ctx.document(), ctx.endOffset())));
+        List<? extends TextEdit> edits = Collections.EMPTY_LIST;
```

Comment:
Nit: might be better to use Collections.emptyList(), as using Collections.EMPTY_LIST leads to an unchecked cast (at least I believe so).